// mongodb/src/cmap/auth/mongodb_oidc/dist/callback_workflow.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
exports.CallbackWorkflow = exports.AUTOMATED_TIMEOUT_MS = exports.HUMAN_TIMEOUT_MS = void 0;
var promises_1 = require("timers/promises");
var error_1 = require("../../../error");
var utils_1 = require("../../../utils");
var command_builders_1 = require("./command_builders");
/** 5 minutes in milliseconds */
exports.HUMAN_TIMEOUT_MS = 300000;
/** 1 minute in milliseconds */
exports.AUTOMATED_TIMEOUT_MS = 60000;
/** Properties allowed on results of callbacks. */
var RESULT_PROPERTIES = ['accessToken', 'expiresInSeconds', 'refreshToken'];
/** Error message when the callback result is invalid. */
var CALLBACK_RESULT_ERROR = 'User provided OIDC callbacks must return a valid object with an accessToken.';
/** The time to throttle callback calls. */
var THROTTLE_MS = 100;
/**
 * OIDC implementation of a callback based workflow.
 * @internal
 */
var CallbackWorkflow = /** @class */ (function () {
    /**
     * Instantiate the callback workflow.
     */
    function CallbackWorkflow(cache, callback) {
        this.cache = cache;
        this.callback = this.withLock(callback);
        this.lastExecutionTime = Date.now() - THROTTLE_MS;
    }
    /**
     * Get the document to add for speculative authentication. This also needs
     * to add a db field from the credentials source.
     */
    CallbackWorkflow.prototype.speculativeAuth = function (connection, credentials) {
        return __awaiter(this, void 0, Promise, function () {
            var accessToken, document;
            return __generator(this, function (_a) {
                // Check if the Client Cache has an access token.
                // If it does, cache the access token in the Connection Cache and send a JwtStepRequest
                // with the cached access token in the speculative authentication SASL payload.
                if (this.cache.hasAccessToken) {
                    accessToken = this.cache.getAccessToken();
                    connection.accessToken = accessToken;
                    document = command_builders_1.finishCommandDocument(accessToken);
                    document.db = credentials.source;
                    return [2 /*return*/, { speculativeAuthenticate: document }];
                }
                return [2 /*return*/, {}];
            });
        });
    };
    /**
     * Reauthenticate the callback workflow. For this we invalidated the access token
     * in the cache and run the authentication steps again. No initial handshake needs
     * to be sent.
     */
    CallbackWorkflow.prototype.reauthenticate = function (connection, credentials) {
        return __awaiter(this, void 0, Promise, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (this.cache.hasAccessToken) {
                            // Reauthentication implies the token has expired.
                            if (connection.accessToken === this.cache.getAccessToken()) {
                                // If connection's access token is the same as the cache's, remove
                                // the token from the cache and connection.
                                this.cache.removeAccessToken();
                                delete connection.accessToken;
                            }
                            else {
                                // If the connection's access token is different from the cache's, set
                                // the cache's token on the connection and do not remove from the
                                // cache.
                                connection.accessToken = this.cache.getAccessToken();
                            }
                        }
                        return [4 /*yield*/, this.execute(connection, credentials)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Starts the callback authentication process. If there is a speculative
     * authentication document from the initial handshake, then we will use that
     * value to get the issuer, otherwise we will send the saslStart command.
     */
    CallbackWorkflow.prototype.startAuthentication = function (connection, credentials, response) {
        return __awaiter(this, void 0, Promise, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(response === null || response === void 0 ? void 0 : response.speculativeAuthenticate)) return [3 /*break*/, 1];
                        result = response.speculativeAuthenticate;
                        return [3 /*break*/, 3];
                    case 1: return [4 /*yield*/, connection.command(utils_1.ns(credentials.source), command_builders_1.startCommandDocument(credentials), undefined)];
                    case 2:
                        result = _a.sent();
                        _a.label = 3;
                    case 3: return [2 /*return*/, result];
                }
            });
        });
    };
    /**
     * Finishes the callback authentication process.
     */
    CallbackWorkflow.prototype.finishAuthentication = function (connection, credentials, token, conversationId) {
        return __awaiter(this, void 0, Promise, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, connection.command(utils_1.ns(credentials.source), command_builders_1.finishCommandDocument(token, conversationId), undefined)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Executes the callback and validates the output.
     */
    CallbackWorkflow.prototype.executeAndValidateCallback = function (params) {
        return __awaiter(this, void 0, Promise, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.callback(params)];
                    case 1:
                        result = _a.sent();
                        // Validate that the result returned by the callback is acceptable. If it is not
                        // we must clear the token result from the cache.
                        if (isCallbackResultInvalid(result)) {
                            throw new error_1.MongoMissingCredentialsError(CALLBACK_RESULT_ERROR);
                        }
                        return [2 /*return*/, result];
                }
            });
        });
    };
    /**
     * Ensure the callback is only executed one at a time and throttles the calls
     * to every 100ms.
     */
    CallbackWorkflow.prototype.withLock = function (callback) {
        var _this = this;
        var lock = Promise.resolve();
        return function (params) { return __awaiter(_this, void 0, Promise, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: 
                    // We do this to ensure that we would never return the result of the
                    // previous lock, only the current callback's value would get returned.
                    return [4 /*yield*/, lock];
                    case 1:
                        // We do this to ensure that we would never return the result of the
                        // previous lock, only the current callback's value would get returned.
                        _a.sent();
                        lock = lock["catch"](function () { return null; })
                            .then(function () { return __awaiter(_this, void 0, void 0, function () {
                            var difference;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        difference = Date.now() - this.lastExecutionTime;
                                        if (!(difference <= THROTTLE_MS)) return [3 /*break*/, 2];
                                        return [4 /*yield*/, promises_1.setTimeout(THROTTLE_MS - difference, { signal: params.timeoutContext })];
                                    case 1:
                                        _a.sent();
                                        _a.label = 2;
                                    case 2:
                                        this.lastExecutionTime = Date.now();
                                        return [4 /*yield*/, callback(params)];
                                    case 3: return [2 /*return*/, _a.sent()];
                                }
                            });
                        }); });
                        return [4 /*yield*/, lock];
                    case 2: return [2 /*return*/, _a.sent()];
                }
            });
        }); };
    };
    return CallbackWorkflow;
}());
exports.CallbackWorkflow = CallbackWorkflow;
/**
 * Determines if a result returned from a request or refresh callback
 * function is invalid. This means the result is nullish, doesn't contain
 * the accessToken required field, and does not contain extra fields.
 */
function isCallbackResultInvalid(tokenResult) {
    if (tokenResult == null || typeof tokenResult !== 'object')
        return true;
    if (!('accessToken' in tokenResult))
        return true;
    return !Object.getOwnPropertyNames(tokenResult).every(function (prop) { return RESULT_PROPERTIES.includes(prop); });
}
