// mongoose/lib/error/dist/setOptionError.dev.js
/*!
 * Module requirements
 */
'use strict';

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var MongooseError = require('./mongooseError');

var util = require('util');

var combinePathErrors = require('../helpers/error/combinePathErrors');
/**
 * Mongoose.set Error
 *
 * @api private
 * @inherits MongooseError
 */


var SetOptionError =
/*#__PURE__*/
function (_MongooseError) {
  _inherits(SetOptionError, _MongooseError);

  function SetOptionError() {
    var _this;

    _classCallCheck(this, SetOptionError);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(SetOptionError).call(this, ''));
    _this.errors = {};
    return _this;
  }
  /**
   * Console.log helper
   */


  _createClass(SetOptionError, [{
    key: "toString",
    value: function toString() {
      return combinePathErrors(this);
    }
    /**
     * inspect helper
     * @api private
     */

  }, {
    key: "inspect",
    value: function inspect() {
      return Object.assign(new Error(this.message), this);
    }
    /**
    * add message
    * @param {String} key
    * @param {String|Error} error
    * @api private
    */

  }, {
    key: "addError",
    value: function addError(key, error) {
      if (error instanceof SetOptionError) {
        var errors = error.errors;

        for (var _i = 0, _Object$keys = Object.keys(errors); _i < _Object$keys.length; _i++) {
          var optionKey = _Object$keys[_i];
          this.addError(optionKey, errors[optionKey]);
        }

        return;
      }

      this.errors[key] = error;
      this.message = combinePathErrors(this);
    }
  }]);

  return SetOptionError;
}(MongooseError);

if (util.inspect.custom) {
  // Avoid Node deprecation warning DEP0079
  SetOptionError.prototype[util.inspect.custom] = SetOptionError.prototype.inspect;
}
/**
 * Helper for JSON.stringify
 * Ensure `name` and `message` show up in toJSON output re: gh-9847
 * @api private
 */


Object.defineProperty(SetOptionError.prototype, 'toJSON', {
  enumerable: false,
  writable: false,
  configurable: true,
  value: function value() {
    return Object.assign({}, this, {
      name: this.name,
      message: this.message
    });
  }
});
Object.defineProperty(SetOptionError.prototype, 'name', {
  value: 'SetOptionError'
});

var SetOptionInnerError =
/*#__PURE__*/
function (_MongooseError2) {
  _inherits(SetOptionInnerError, _MongooseError2);

  /**
   * Error for the "errors" array in "SetOptionError" with consistent message
   * @param {String} key
   */
  function SetOptionInnerError(key) {
    _classCallCheck(this, SetOptionInnerError);

    return _possibleConstructorReturn(this, _getPrototypeOf(SetOptionInnerError).call(this, "\"".concat(key, "\" is not a valid option to set")));
  }

  return SetOptionInnerError;
}(MongooseError);

SetOptionError.SetOptionInnerError = SetOptionInnerError;
/*!
 * Module exports
 */

module.exports = SetOptionError;
