// mongodb/src/cmap/dist/metrics.js
"use strict";
exports.__esModule = true;
exports.ConnectionPoolMetrics = void 0;
/** @internal */
var ConnectionPoolMetrics = /** @class */ (function () {
    function ConnectionPoolMetrics() {
        this.txnConnections = 0;
        this.cursorConnections = 0;
        this.otherConnections = 0;
    }
    /**
     * Mark a connection as pinned for a specific operation.
     */
    ConnectionPoolMetrics.prototype.markPinned = function (pinType) {
        if (pinType === ConnectionPoolMetrics.TXN) {
            this.txnConnections += 1;
        }
        else if (pinType === ConnectionPoolMetrics.CURSOR) {
            this.cursorConnections += 1;
        }
        else {
            this.otherConnections += 1;
        }
    };
    /**
     * Unmark a connection as pinned for an operation.
     */
    ConnectionPoolMetrics.prototype.markUnpinned = function (pinType) {
        if (pinType === ConnectionPoolMetrics.TXN) {
            this.txnConnections -= 1;
        }
        else if (pinType === ConnectionPoolMetrics.CURSOR) {
            this.cursorConnections -= 1;
        }
        else {
            this.otherConnections -= 1;
        }
    };
    /**
     * Return information about the cmap metrics as a string.
     */
    ConnectionPoolMetrics.prototype.info = function (maxPoolSize) {
        return ('Timed out while checking out a connection from connection pool: ' +
            ("maxPoolSize: " + maxPoolSize + ", ") +
            ("connections in use by cursors: " + this.cursorConnections + ", ") +
            ("connections in use by transactions: " + this.txnConnections + ", ") +
            ("connections in use by other operations: " + this.otherConnections));
    };
    /**
     * Reset the metrics to the initial values.
     */
    ConnectionPoolMetrics.prototype.reset = function () {
        this.txnConnections = 0;
        this.cursorConnections = 0;
        this.otherConnections = 0;
    };
    ConnectionPoolMetrics.TXN = 'txn';
    ConnectionPoolMetrics.CURSOR = 'cursor';
    ConnectionPoolMetrics.OTHER = 'other';
    return ConnectionPoolMetrics;
}());
exports.ConnectionPoolMetrics = ConnectionPoolMetrics;
