// mongodb/src/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
exports.__esModule = true;
exports.UnorderedBulkOperation = exports.OrderedBulkOperation = exports.MongoClient = exports.ListIndexesCursor = exports.ListCollectionsCursor = exports.GridFSBucketWriteStream = exports.GridFSBucketReadStream = exports.GridFSBucket = exports.FindCursor = exports.Db = exports.Collection = exports.ClientSession = exports.ChangeStream = exports.CancellationToken = exports.AggregationCursor = exports.Admin = exports.AbstractCursor = void 0;
var admin_1 = require("./admin");
exports.Admin = admin_1.Admin;
var ordered_1 = require("./bulk/ordered");
exports.OrderedBulkOperation = ordered_1.OrderedBulkOperation;
var unordered_1 = require("./bulk/unordered");
exports.UnorderedBulkOperation = unordered_1.UnorderedBulkOperation;
var change_stream_1 = require("./change_stream");
exports.ChangeStream = change_stream_1.ChangeStream;
var collection_1 = require("./collection");
exports.Collection = collection_1.Collection;
var abstract_cursor_1 = require("./cursor/abstract_cursor");
exports.AbstractCursor = abstract_cursor_1.AbstractCursor;
var aggregation_cursor_1 = require("./cursor/aggregation_cursor");
exports.AggregationCursor = aggregation_cursor_1.AggregationCursor;
var find_cursor_1 = require("./cursor/find_cursor");
exports.FindCursor = find_cursor_1.FindCursor;
var list_collections_cursor_1 = require("./cursor/list_collections_cursor");
exports.ListCollectionsCursor = list_collections_cursor_1.ListCollectionsCursor;
var list_indexes_cursor_1 = require("./cursor/list_indexes_cursor");
exports.ListIndexesCursor = list_indexes_cursor_1.ListIndexesCursor;
var db_1 = require("./db");
exports.Db = db_1.Db;
var gridfs_1 = require("./gridfs");
exports.GridFSBucket = gridfs_1.GridFSBucket;
var download_1 = require("./gridfs/download");
exports.GridFSBucketReadStream = download_1.GridFSBucketReadStream;
var upload_1 = require("./gridfs/upload");
exports.GridFSBucketWriteStream = upload_1.GridFSBucketWriteStream;
var mongo_client_1 = require("./mongo_client");
exports.MongoClient = mongo_client_1.MongoClient;
var mongo_types_1 = require("./mongo_types");
exports.CancellationToken = mongo_types_1.CancellationToken;
var sessions_1 = require("./sessions");
exports.ClientSession = sessions_1.ClientSession;
/** @public */
var bson_1 = require("./bson");
__createBinding(exports, bson_1, "BSON");
var bson_2 = require("./bson");
__createBinding(exports, bson_2, "Binary");
__createBinding(exports, bson_2, "BSONRegExp");
__createBinding(exports, bson_2, "BSONSymbol");
__createBinding(exports, bson_2, "BSONType");
__createBinding(exports, bson_2, "Code");
__createBinding(exports, bson_2, "DBRef");
__createBinding(exports, bson_2, "Decimal128");
__createBinding(exports, bson_2, "Double");
__createBinding(exports, bson_2, "Int32");
__createBinding(exports, bson_2, "Long");
__createBinding(exports, bson_2, "MaxKey");
__createBinding(exports, bson_2, "MinKey");
__createBinding(exports, bson_2, "ObjectId");
__createBinding(exports, bson_2, "Timestamp");
__createBinding(exports, bson_2, "UUID");
var common_1 = require("./bulk/common");
__createBinding(exports, common_1, "MongoBulkWriteError");
var client_encryption_1 = require("./client-side-encryption/client_encryption");
__createBinding(exports, client_encryption_1, "ClientEncryption");
var change_stream_cursor_1 = require("./cursor/change_stream_cursor");
__createBinding(exports, change_stream_cursor_1, "ChangeStreamCursor");
var error_1 = require("./error");
__createBinding(exports, error_1, "MongoAPIError");
__createBinding(exports, error_1, "MongoAWSError");
__createBinding(exports, error_1, "MongoAzureError");
__createBinding(exports, error_1, "MongoBatchReExecutionError");
__createBinding(exports, error_1, "MongoChangeStreamError");
__createBinding(exports, error_1, "MongoCompatibilityError");
__createBinding(exports, error_1, "MongoCursorExhaustedError");
__createBinding(exports, error_1, "MongoCursorInUseError");
__createBinding(exports, error_1, "MongoDecompressionError");
__createBinding(exports, error_1, "MongoDriverError");
__createBinding(exports, error_1, "MongoError");
__createBinding(exports, error_1, "MongoExpiredSessionError");
__createBinding(exports, error_1, "MongoGCPError");
__createBinding(exports, error_1, "MongoGridFSChunkError");
__createBinding(exports, error_1, "MongoGridFSStreamError");
__createBinding(exports, error_1, "MongoInvalidArgumentError");
__createBinding(exports, error_1, "MongoKerberosError");
__createBinding(exports, error_1, "MongoMissingCredentialsError");
__createBinding(exports, error_1, "MongoMissingDependencyError");
__createBinding(exports, error_1, "MongoNetworkError");
__createBinding(exports, error_1, "MongoNetworkTimeoutError");
__createBinding(exports, error_1, "MongoNotConnectedError");
__createBinding(exports, error_1, "MongoOIDCError");
__createBinding(exports, error_1, "MongoParseError");
__createBinding(exports, error_1, "MongoRuntimeError");
__createBinding(exports, error_1, "MongoServerClosedError");
__createBinding(exports, error_1, "MongoServerError");
__createBinding(exports, error_1, "MongoServerSelectionError");
__createBinding(exports, error_1, "MongoSystemError");
__createBinding(exports, error_1, "MongoTailableCursorError");
__createBinding(exports, error_1, "MongoTopologyClosedError");
__createBinding(exports, error_1, "MongoTransactionError");
__createBinding(exports, error_1, "MongoUnexpectedServerResponseError");
__createBinding(exports, error_1, "MongoWriteConcernError");
var resource_management_1 = require("./resource_management");
__createBinding(exports, resource_management_1, "configureExplicitResourceManagement");
// enums
var common_2 = require("./bulk/common");
__createBinding(exports, common_2, "BatchType");
var auto_encrypter_1 = require("./client-side-encryption/auto_encrypter");
__createBinding(exports, auto_encrypter_1, "AutoEncryptionLoggerLevel");
var gssapi_1 = require("./cmap/auth/gssapi");
__createBinding(exports, gssapi_1, "GSSAPICanonicalizationValue");
var providers_1 = require("./cmap/auth/providers");
__createBinding(exports, providers_1, "AuthMechanism");
var compression_1 = require("./cmap/wire_protocol/compression");
__createBinding(exports, compression_1, "Compressor");
var abstract_cursor_2 = require("./cursor/abstract_cursor");
__createBinding(exports, abstract_cursor_2, "CURSOR_FLAGS");
var error_2 = require("./error");
__createBinding(exports, error_2, "MongoErrorLabel");
var explain_1 = require("./explain");
__createBinding(exports, explain_1, "ExplainVerbosity");
var mongo_client_2 = require("./mongo_client");
__createBinding(exports, mongo_client_2, "ServerApiVersion");
var find_and_modify_1 = require("./operations/find_and_modify");
__createBinding(exports, find_and_modify_1, "ReturnDocument");
var set_profiling_level_1 = require("./operations/set_profiling_level");
__createBinding(exports, set_profiling_level_1, "ProfilingLevel");
var read_concern_1 = require("./read_concern");
__createBinding(exports, read_concern_1, "ReadConcernLevel");
var read_preference_1 = require("./read_preference");
__createBinding(exports, read_preference_1, "ReadPreferenceMode");
var common_3 = require("./sdam/common");
__createBinding(exports, common_3, "ServerType");
__createBinding(exports, common_3, "TopologyType");
// Helper classes
var read_concern_2 = require("./read_concern");
__createBinding(exports, read_concern_2, "ReadConcern");
var read_preference_2 = require("./read_preference");
__createBinding(exports, read_preference_2, "ReadPreference");
var write_concern_1 = require("./write_concern");
__createBinding(exports, write_concern_1, "WriteConcern");
// events
var command_monitoring_events_1 = require("./cmap/command_monitoring_events");
__createBinding(exports, command_monitoring_events_1, "CommandFailedEvent");
__createBinding(exports, command_monitoring_events_1, "CommandStartedEvent");
__createBinding(exports, command_monitoring_events_1, "CommandSucceededEvent");
var connection_pool_events_1 = require("./cmap/connection_pool_events");
__createBinding(exports, connection_pool_events_1, "ConnectionCheckedInEvent");
__createBinding(exports, connection_pool_events_1, "ConnectionCheckedOutEvent");
__createBinding(exports, connection_pool_events_1, "ConnectionCheckOutFailedEvent");
__createBinding(exports, connection_pool_events_1, "ConnectionCheckOutStartedEvent");
__createBinding(exports, connection_pool_events_1, "ConnectionClosedEvent");
__createBinding(exports, connection_pool_events_1, "ConnectionCreatedEvent");
__createBinding(exports, connection_pool_events_1, "ConnectionPoolClearedEvent");
__createBinding(exports, connection_pool_events_1, "ConnectionPoolClosedEvent");
__createBinding(exports, connection_pool_events_1, "ConnectionPoolCreatedEvent");
__createBinding(exports, connection_pool_events_1, "ConnectionPoolMonitoringEvent");
__createBinding(exports, connection_pool_events_1, "ConnectionPoolReadyEvent");
__createBinding(exports, connection_pool_events_1, "ConnectionReadyEvent");
var events_1 = require("./sdam/events");
__createBinding(exports, events_1, "ServerClosedEvent");
__createBinding(exports, events_1, "ServerDescriptionChangedEvent");
__createBinding(exports, events_1, "ServerHeartbeatFailedEvent");
__createBinding(exports, events_1, "ServerHeartbeatStartedEvent");
__createBinding(exports, events_1, "ServerHeartbeatSucceededEvent");
__createBinding(exports, events_1, "ServerOpeningEvent");
__createBinding(exports, events_1, "TopologyClosedEvent");
__createBinding(exports, events_1, "TopologyDescriptionChangedEvent");
__createBinding(exports, events_1, "TopologyOpeningEvent");
var server_selection_events_1 = require("./sdam/server_selection_events");
__createBinding(exports, server_selection_events_1, "ServerSelectionEvent");
__createBinding(exports, server_selection_events_1, "ServerSelectionFailedEvent");
__createBinding(exports, server_selection_events_1, "ServerSelectionStartedEvent");
__createBinding(exports, server_selection_events_1, "ServerSelectionSucceededEvent");
__createBinding(exports, server_selection_events_1, "WaitingForSuitableServerEvent");
var srv_polling_1 = require("./sdam/srv_polling");
__createBinding(exports, srv_polling_1, "SrvPollingEvent");
var errors_1 = require("./client-side-encryption/errors");
__createBinding(exports, errors_1, "MongoCryptAzureKMSRequestError");
__createBinding(exports, errors_1, "MongoCryptCreateDataKeyError");
__createBinding(exports, errors_1, "MongoCryptCreateEncryptedCollectionError");
__createBinding(exports, errors_1, "MongoCryptError");
__createBinding(exports, errors_1, "MongoCryptInvalidArgumentError");
__createBinding(exports, errors_1, "MongoCryptKMSRequestNetworkTimeoutError");
var mongo_client_auth_providers_1 = require("./mongo_client_auth_providers");
__createBinding(exports, mongo_client_auth_providers_1, "MongoClientAuthProviders");
