// mongodb/src/dist/read_concern.js
"use strict";
exports.__esModule = true;
exports.ReadConcern = exports.ReadConcernLevel = void 0;
/** @public */
exports.ReadConcernLevel = Object.freeze({
    local: 'local',
    majority: 'majority',
    linearizable: 'linearizable',
    available: 'available',
    snapshot: 'snapshot'
});
/**
 * The MongoDB ReadConcern, which allows for control of the consistency and isolation properties
 * of the data read from replica sets and replica set shards.
 * @public
 *
 * @see https://www.mongodb.com/docs/manual/reference/read-concern/index.html
 */
var ReadConcern = /** @class */ (function () {
    /** Constructs a ReadConcern from the read concern level.*/
    function ReadConcern(level) {
        var _a;
        /**
         * A spec test exists that allows level to be any string.
         * "invalid readConcern with out stage"
         * @see ./test/spec/crud/v2/aggregate-out-readConcern.json
         * @see https://github.com/mongodb/specifications/blob/master/source/read-write-concern/read-write-concern.rst#unknown-levels-and-additional-options-for-string-based-readconcerns
         */
        this.level = (_a = exports.ReadConcernLevel[level]) !== null && _a !== void 0 ? _a : level;
    }
    /**
     * Construct a ReadConcern given an options object.
     *
     * @param options - The options object from which to extract the write concern.
     */
    ReadConcern.fromOptions = function (options) {
        if (options == null) {
            return;
        }
        if (options.readConcern) {
            var readConcern = options.readConcern;
            if (readConcern instanceof ReadConcern) {
                return readConcern;
            }
            else if (typeof readConcern === 'string') {
                return new ReadConcern(readConcern);
            }
            else if ('level' in readConcern && readConcern.level) {
                return new ReadConcern(readConcern.level);
            }
        }
        if (options.level) {
            return new ReadConcern(options.level);
        }
        return;
    };
    Object.defineProperty(ReadConcern, "MAJORITY", {
        get: function () {
            return exports.ReadConcernLevel.majority;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ReadConcern, "AVAILABLE", {
        get: function () {
            return exports.ReadConcernLevel.available;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ReadConcern, "LINEARIZABLE", {
        get: function () {
            return exports.ReadConcernLevel.linearizable;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ReadConcern, "SNAPSHOT", {
        get: function () {
            return exports.ReadConcernLevel.snapshot;
        },
        enumerable: false,
        configurable: true
    });
    ReadConcern.prototype.toJSON = function () {
        return { level: this.level };
    };
    return ReadConcern;
}());
exports.ReadConcern = ReadConcern;
