// mongodb/lib/dist/explain.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Explain = exports.ExplainVerbosity = void 0;
/** @public */

exports.ExplainVerbosity = Object.freeze({
  queryPlanner: 'queryPlanner',
  queryPlannerExtended: 'queryPlannerExtended',
  executionStats: 'executionStats',
  allPlansExecution: 'allPlansExecution'
});
/** @internal */

var Explain =
/*#__PURE__*/
function () {
  function Explain(verbosity, maxTimeMS) {
    _classCallCheck(this, Explain);

    if (typeof verbosity === 'boolean') {
      this.verbosity = verbosity ? exports.ExplainVerbosity.allPlansExecution : exports.ExplainVerbosity.queryPlanner;
    } else {
      this.verbosity = verbosity;
    }

    this.maxTimeMS = maxTimeMS;
  }

  _createClass(Explain, null, [{
    key: "fromOptions",
    value: function fromOptions() {
      var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
          explain = _ref.explain;

      if (explain == null) return;

      if (typeof explain === 'boolean' || typeof explain === 'string') {
        return new Explain(explain);
      }

      var verbosity = explain.verbosity,
          maxTimeMS = explain.maxTimeMS;
      return new Explain(verbosity, maxTimeMS);
    }
  }]);

  return Explain;
}();

exports.Explain = Explain;
