// sift/src/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
exports.__esModule = true;
exports.createQueryOperation = exports.createEqualsOperation = exports.createDefaultQueryOperation = exports.createOperationTester = exports.createQueryTester = exports.EqualsOperation = void 0;
var defaultOperations = require("./operations");
var core_1 = require("./core");
exports.createQueryTester = core_1.createQueryTester;
exports.EqualsOperation = core_1.EqualsOperation;
exports.createQueryOperation = core_1.createQueryOperation;
exports.createEqualsOperation = core_1.createEqualsOperation;
exports.createOperationTester = core_1.createOperationTester;
var createDefaultQueryOperation = function (query, ownerQuery, _a) {
    var _b = _a === void 0 ? {} : _a, compare = _b.compare, operations = _b.operations;
    return core_1.createQueryOperation(query, ownerQuery, {
        compare: compare,
        operations: Object.assign({}, defaultOperations, operations || {})
    });
};
exports.createDefaultQueryOperation = createDefaultQueryOperation;
var createDefaultQueryTester = function (query, options) {
    if (options === void 0) { options = {}; }
    var op = createDefaultQueryOperation(query, null, options);
    return core_1.createOperationTester(op);
};
__exportStar(require("./operations"), exports);
exports["default"] = createDefaultQueryTester;
