// mongodb/src/cmap/auth/mongodb_oidc/dist/azure_machine_workflow.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
exports.AzureMachineWorkflow = void 0;
var azure_1 = require("../../../client-side-encryption/providers/azure");
var error_1 = require("../../../error");
var utils_1 = require("../../../utils");
var machine_workflow_1 = require("./machine_workflow");
/** Azure request headers. */
var AZURE_HEADERS = Object.freeze({ Metadata: 'true', Accept: 'application/json' });
/** Invalid endpoint result error. */
var ENDPOINT_RESULT_ERROR = 'Azure endpoint did not return a value with only access_token and expires_in properties';
/** Error for when the token audience is missing in the environment. */
var TOKEN_RESOURCE_MISSING_ERROR = 'TOKEN_RESOURCE must be set in the auth mechanism properties when ENVIRONMENT is azure.';
/**
 * Device workflow implementation for Azure.
 *
 * @internal
 */
var AzureMachineWorkflow = /** @class */ (function (_super) {
    __extends(AzureMachineWorkflow, _super);
    /**
     * Instantiate the machine workflow.
     */
    function AzureMachineWorkflow(cache) {
        return _super.call(this, cache) || this;
    }
    /**
     * Get the token from the environment.
     */
    AzureMachineWorkflow.prototype.getToken = function (credentials) {
        return __awaiter(this, void 0, Promise, function () {
            var tokenAudience, username, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        tokenAudience = credentials === null || credentials === void 0 ? void 0 : credentials.mechanismProperties.TOKEN_RESOURCE;
                        username = credentials === null || credentials === void 0 ? void 0 : credentials.username;
                        if (!tokenAudience) {
                            throw new error_1.MongoAzureError(TOKEN_RESOURCE_MISSING_ERROR);
                        }
                        return [4 /*yield*/, getAzureTokenData(tokenAudience, username)];
                    case 1:
                        response = _a.sent();
                        if (!isEndpointResultValid(response)) {
                            throw new error_1.MongoAzureError(ENDPOINT_RESULT_ERROR);
                        }
                        return [2 /*return*/, response];
                }
            });
        });
    };
    return AzureMachineWorkflow;
}(machine_workflow_1.MachineWorkflow));
exports.AzureMachineWorkflow = AzureMachineWorkflow;
/**
 * Hit the Azure endpoint to get the token data.
 */
function getAzureTokenData(tokenAudience, username) {
    return __awaiter(this, void 0, Promise, function () {
        var url, response, result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    url = new URL(azure_1.AZURE_BASE_URL);
                    azure_1.addAzureParams(url, tokenAudience, username);
                    return [4 /*yield*/, utils_1.get(url, {
                            headers: AZURE_HEADERS
                        })];
                case 1:
                    response = _a.sent();
                    if (response.status !== 200) {
                        throw new error_1.MongoAzureError("Status code " + response.status + " returned from the Azure endpoint. Response body: " + response.body);
                    }
                    result = JSON.parse(response.body);
                    return [2 /*return*/, {
                            access_token: result.access_token,
                            expires_in: Number(result.expires_in)
                        }];
            }
        });
    });
}
/**
 * Determines if a result returned from the endpoint is valid.
 * This means the result is not nullish, contains the access_token required field
 * and the expires_in required field.
 */
function isEndpointResultValid(token) {
    if (token == null || typeof token !== 'object')
        return false;
    return ('access_token' in token &&
        typeof token.access_token === 'string' &&
        'expires_in' in token &&
        typeof token.expires_in === 'number');
}
