// mongodb/src/dist/write_concern.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
exports.__esModule = true;
exports.throwIfWriteConcernError = exports.WriteConcern = exports.WRITE_CONCERN_KEYS = void 0;
var responses_1 = require("./cmap/wire_protocol/responses");
var error_1 = require("./error");
exports.WRITE_CONCERN_KEYS = ['w', 'wtimeout', 'j', 'journal', 'fsync'];
/**
 * A MongoDB WriteConcern, which describes the level of acknowledgement
 * requested from MongoDB for write operations.
 * @public
 *
 * @see https://www.mongodb.com/docs/manual/reference/write-concern/
 */
var WriteConcern = /** @class */ (function () {
    /**
     * Constructs a WriteConcern from the write concern properties.
     * @param w - request acknowledgment that the write operation has propagated to a specified number of mongod instances or to mongod instances with specified tags.
     * @param wtimeoutMS - specify a time limit to prevent write operations from blocking indefinitely
     * @param journal - request acknowledgment that the write operation has been written to the on-disk journal
     * @param fsync - equivalent to the j option. Is deprecated and will be removed in the next major version.
     */
    function WriteConcern(w, wtimeoutMS, journal, fsync) {
        if (w != null) {
            if (!Number.isNaN(Number(w))) {
                this.w = Number(w);
            }
            else {
                this.w = w;
            }
        }
        if (wtimeoutMS != null) {
            this.wtimeoutMS = this.wtimeout = wtimeoutMS;
        }
        if (journal != null) {
            this.journal = this.j = journal;
        }
        if (fsync != null) {
            this.journal = this.j = fsync ? true : false;
        }
    }
    /**
     * Apply a write concern to a command document. Will modify and return the command.
     */
    WriteConcern.apply = function (command, writeConcern) {
        var wc = {};
        // The write concern document sent to the server has w/wtimeout/j fields.
        if (writeConcern.w != null)
            wc.w = writeConcern.w;
        if (writeConcern.wtimeoutMS != null)
            wc.wtimeout = writeConcern.wtimeoutMS;
        if (writeConcern.journal != null)
            wc.j = writeConcern.j;
        command.writeConcern = wc;
        return command;
    };
    /** Construct a WriteConcern given an options object. */
    WriteConcern.fromOptions = function (options, inherit) {
        if (options == null)
            return undefined;
        inherit = inherit !== null && inherit !== void 0 ? inherit : {};
        var opts;
        if (typeof options === 'string' || typeof options === 'number') {
            opts = { w: options };
        }
        else if (options instanceof WriteConcern) {
            opts = options;
        }
        else {
            opts = options.writeConcern;
        }
        var parentOpts = inherit instanceof WriteConcern ? inherit : inherit.writeConcern;
        var _a = __assign(__assign({}, parentOpts), opts), _b = _a.w, w = _b === void 0 ? undefined : _b, _c = _a.wtimeout, wtimeout = _c === void 0 ? undefined : _c, _d = _a.j, j = _d === void 0 ? undefined : _d, _e = _a.fsync, fsync = _e === void 0 ? undefined : _e, _f = _a.journal, journal = _f === void 0 ? undefined : _f, _g = _a.wtimeoutMS, wtimeoutMS = _g === void 0 ? undefined : _g;
        if (w != null ||
            wtimeout != null ||
            wtimeoutMS != null ||
            j != null ||
            journal != null ||
            fsync != null) {
            return new WriteConcern(w, wtimeout !== null && wtimeout !== void 0 ? wtimeout : wtimeoutMS, j !== null && j !== void 0 ? j : journal, fsync);
        }
        return undefined;
    };
    return WriteConcern;
}());
exports.WriteConcern = WriteConcern;
/** Called with either a plain object or MongoDBResponse */
function throwIfWriteConcernError(response) {
    if (typeof response === 'object' && response != null) {
        var writeConcernError = responses_1.MongoDBResponse.is(response) && response.has('writeConcernError')
            ? response.toObject()
            : !responses_1.MongoDBResponse.is(response) && 'writeConcernError' in response
                ? response
                : null;
        if (writeConcernError != null) {
            throw new error_1.MongoWriteConcernError(writeConcernError);
        }
    }
}
exports.throwIfWriteConcernError = throwIfWriteConcernError;
