// mongodb/src/cmap/dist/stream_description.js
"use strict";
exports.__esModule = true;
exports.StreamDescription = void 0;
var bson_1 = require("../bson");
var common_1 = require("../sdam/common");
var server_description_1 = require("../sdam/server_description");
var RESPONSE_FIELDS = [
    'minWireVersion',
    'maxWireVersion',
    'maxBsonObjectSize',
    'maxMessageSizeBytes',
    'maxWriteBatchSize',
    'logicalSessionTimeoutMinutes'
];
/** @public */
var StreamDescription = /** @class */ (function () {
    function StreamDescription(address, options) {
        this.hello = null;
        this.address = address;
        this.type = common_1.ServerType.Unknown;
        this.minWireVersion = undefined;
        this.maxWireVersion = undefined;
        this.maxBsonObjectSize = 16777216;
        this.maxMessageSizeBytes = 48000000;
        this.maxWriteBatchSize = 100000;
        this.logicalSessionTimeoutMinutes = options === null || options === void 0 ? void 0 : options.logicalSessionTimeoutMinutes;
        this.loadBalanced = !!(options === null || options === void 0 ? void 0 : options.loadBalanced);
        this.compressors =
            options && options.compressors && Array.isArray(options.compressors)
                ? options.compressors
                : [];
        this.serverConnectionId = null;
    }
    StreamDescription.prototype.receiveResponse = function (response) {
        if (response == null) {
            return;
        }
        this.hello = response;
        this.type = server_description_1.parseServerType(response);
        if ('connectionId' in response) {
            this.serverConnectionId = this.parseServerConnectionID(response.connectionId);
        }
        else {
            this.serverConnectionId = null;
        }
        for (var _i = 0, RESPONSE_FIELDS_1 = RESPONSE_FIELDS; _i < RESPONSE_FIELDS_1.length; _i++) {
            var field = RESPONSE_FIELDS_1[_i];
            if (response[field] != null) {
                this[field] = response[field];
            }
            // testing case
            if ('__nodejs_mock_server__' in response) {
                this.__nodejs_mock_server__ = response['__nodejs_mock_server__'];
            }
        }
        if (response.compression) {
            this.compressor = this.compressors.filter(function (c) { var _a; return (_a = response.compression) === null || _a === void 0 ? void 0 : _a.includes(c); })[0];
        }
    };
    /* @internal */
    StreamDescription.prototype.parseServerConnectionID = function (serverConnectionId) {
        // Connection ids are always integral, so it's safe to coerce doubles as well as
        // any integral type.
        return bson_1.Long.isLong(serverConnectionId)
            ? serverConnectionId.toBigInt()
            : // @ts-expect-error: Doubles are coercible to number
                BigInt(serverConnectionId);
    };
    return StreamDescription;
}());
exports.StreamDescription = StreamDescription;
